src/smt/theory_str_noodler/parikh_image.h

```diff
@@ -0,0 +1,217 @@
+#ifndef Z3_STR_PARIKH_H_
+#define Z3_STR_PARIKH_H_
+
+#include <iostream>
+#include <map>
+#include <set>
+#include <queue>
+#include <string>
+#include <memory>
+#include <concepts>
+#include <compare>
+
+#include <mata/nfa/nfa.hh>
+#include <mata/nfa/strings.hh>
+#include <mata/nfa/builder.hh>
+
+#include "formula.h"
+#include "counter_automaton.h"
+#include "formula_preprocess.h"
+
+namespace smt::noodler::parikh {
+
+using Transition = std::tuple<mata::nfa::State, mata::Symbol, mata::nfa::State>;
```

Comment:
Wouldn't it better to use mata's version?

Reply:
Mata doesn't allow to compare transitions, which is necessary here. If we add comparison to mata I can switch it.

Reply:
Parikh image is not used anywhere else right now, right? If not, maybe I would move the stuff related to Parikh image (probably just the files `parikh_image.h` and `parikh_image.cpp`) to its own PR and we can sort all this later.

Reply:
It does not make sense as the content of this PR depends heavily on Parikh image computation. After merge we can discuss what we could do with it. Putting it to mata is a bad idea imo.

Reply:
Ah ok, from the description of PR it seemed that you didn't use Parikh image at the end.

Reply:
And I only meant putting comparison of transitions in Mata, so you do not have to use `Transition`.

Reply:
Yes, I am planning to do the transition comparison in mata.